Implement retroactive IS locking of transactions first seen in blocks instead of mempool

codablock opened the pull request:
This implements "retroactive signing" of TXs which were not known before a block appears that contains the TX. This implicitly enables retroactive signing of CLSIGs when included transactions were previously considered "unsafe".

The PR also includes a few fixes for issues I encountered while implementing/testing this.

UdjinM6 commented on src/llmq/quorums_chainlocks.cpp:
```diff
@@ -332,12 +332,10 @@ void CChainLocksHandler::NewPoWValidBlock(const CBlockIndex* pindex, const std::
-                tx->nType == TRANSACTION_QUORUM_COMMITMENT) {
-                continue;
-            }
+    for (size_t i = 0; i < block->vtx.size(); i++) {
```
would prefer
```
for (const auto& tx : block->vtx) {
    if (tx->IsCoinBase() || tx->vin.empty()) {
```

codablock replied:
Agree, looks better. Fixed.

UdjinM6 commented on src/llmq/quorums_chainlocks.cpp:
```diff
@@ -347,11 +345,8 @@ void CChainLocksHandler::NewPoWValidBlock(const CBlockIndex* pindex, const std::
-            tx.nType == TRANSACTION_QUORUM_COMMITMENT) {
-            return;
-        }
+    if (posInBlock == 0 || tx.vin.empty()) {
```
same here
```
if (tx.IsCoinBase() || tx.vin.empty()) {
```

UdjinM6 commented on src/llmq/quorums_instantsend.cpp:
```diff
@@ -843,7 +843,11 @@ void CInstantSendManager::RetryLockTxs(const uint256& lockedParentTx)
+    // scan blocks until we hit the last chainlocked block we know of. Also stop scanning after a depth of 6 to avoid
+    // signing thousands of TXs at once. Also, after a depth of 6, blocks get eligible for ChainLocking even if unsafe
+    // TXs are included, so there is no need to retroactively sign these.
+    int depth = 6;
```
```suggestion
    int depth = 0;
```

codablock replied:
Whoops :see_no_evil: This one was silly...and I'm actually trying to figure out since 10 minutes what's wrong with tests :D

codablock commented:
Please wait with merging, I'd like to do a rebase first after #2774 is merged (I hope this fixes test failures here)

codablock commented:
Rebased this PR on #2776 and added a fix for test failures. Will rebase on develop after #2776 is merged.

codablock commented:
Tests are finally green, ready for merge after re-ACK

UdjinM6 commented:
I'm still not sure if I like the idea tbh 🙈 

Also, can we split this PR into two - one for the fixes and the other one for the actual retroactive signing implementation?

codablock commented:
Ok, I'll do some PR splitting tomorrow.

codablock commented:
@UdjinM6 PR with only fixes/refactorings created in #2779

codablock commented:
I've also rebased this PR on top of #2779 so that the ordering and conflict resolution is already in-place. I also added a commit on top to fix an issue with NewPoWValidBlock

codablock commented:
Rebased on develop and force-pushed. PR now only contains commits related to the PR.

UdjinM6 commented on src/llmq/quorums_instantsend.cpp:
```diff
@@ -820,6 +820,39 @@ void CInstantSendManager::RetryLockMempoolTxs(const uint256& lockedParentTx)
+    const CBlockIndex* pindexWalk = nullptr;
+    {
+        LOCK(cs);
+        lastChainLockBlock = db.GetLastChainLockBlock();
```
can bail out early here
```
if (lastChainLockBlock.IsNull()) {
    return;
}
```

codablock replied:
Not really, as we might have entries in `txs` from the mempool loop at the top.

codablock commented on src/llmq/quorums_instantsend.cpp:
```diff
@@ -708,8 +711,17 @@ void CInstantSendManager::SyncTransaction(const CTransaction& tx, const CBlockIn
 
-    if (IsLocked(tx.GetHash())) {
-        RetryLockMempoolTxs(tx.GetHash());
+    if (posInBlock == 0) {
```
done, and also added a check for vin.empty()
